lib/spack/spack/mirrors/filter.py

```diff
@@ -0,0 +1,38 @@
+# Copyright Spack Project Developers. See COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+from typing import List
+
+import spack.spec
+from spack.mirrors.mirror import Mirror
+
+
+class MirrorSpecFilter:
+    def __init__(self, mirror: Mirror):
+        self.exclude = [spack.spec.Spec(spec) for spec in mirror.exclusions]
+        self.include = [spack.spec.Spec(spec) for spec in mirror.inclusions]
+
+    def __call__(self, specs: List[spack.spec.Spec]):
+        """
+        Determine the intersection of include/exclude filters
+        Tie goes to keeping
+
+        skip  | keep  | outcome
+        ------------------------
+        False | False | Keep
+        True  | True  | Keep
```

Comment:
I think this is okay, but I think calling the second field include is kind of strange since things that don't match the include are also included. Adding a `@:` match for include disables the exclude, which makes the meaning of exclude kind of wonky.

Really this is a exclude, and override/exempt from exclude, which is a little different than include. Maybe for each exclusion there could be an optional exemptions field?

```yaml
exclude:
    - match: dev_path=*
      exempt:
          - zlib
          - ~shared
    - intel-oneapi-runtime
```

Reply:
just be consistent with views/modules etc. https://spack.readthedocs.io/en/latest/environments.html#advanced-view-configuration

Reply:
The main issue with that is to get detailed exclusions in a dev environment, where the majority of dev specs should be excluded, is you need to use `<package_name> dev_path=*` to say you want to push one spec to a shared developer build cache. That is potentially a lot of config if there are 10s of dev specs in the environment and leads to more potential for error.

Granted, a tool like `spack-manager` could maintain a reference block in `mirrors.yaml` that managed the full list of develop specs maybe that could work...

```yaml
.spack-manager:
  develop-specs: &sm_mirror_exclusions
    - zlib dev_path=*
    - ncurses dev_path=*

mirrors:
  my_mirror:
     select: "@:"
     exclude:
     - *sm_mirror_exclusions
     - intel-oneapi-runtime
```

Reply:
same problem for modules and views...

Reply:
https://spack.readthedocs.io/en/latest/module_file_support.html#exclude-or-include-specific-module-files looks like views already disagree, the module way is more sensible?

Reply:
@kwryankrattiger I added the `--exclude-from-file` to `spack mirror add` so complex exclusion configurations can be supported.  I __think__ what I have matches modules. I agree exemption is likely a better name. I can make that change here, but it is adding another bifurcation of syntax.